octopoes/octopoes/models/ooi/web.py

```diff
@@ -284,3 +284,51 @@ def format_reference_human_readable(cls, reference: Reference) -> str:
         api_url = format_web_url_token(t.rest_api.api_url)
 
         return f"{rule} @ {api_url}"
+
+
+class Cookie(OOI):
+    object_type: Literal["Cookie"] = "Cookie"
+
+    name: str
+    value: str
+    expires: Optional[str]
+    max_age: Optional[int]
+    # https://datatracker.ietf.org/doc/html/rfc6265#section-5.3 p6
+    domain: Reference = ReferenceField(Hostname)
+    # https://datatracker.ietf.org/doc/html/rfc6265#section-5.3 p7
+    path: Optional[str]
+    created: Optional[str]
+    # last-access-time: # not used in openkat context
+    # https://datatracker.ietf.org/doc/html/rfc6265#section-5.3 p3
+    persistent: bool
+    # https://datatracker.ietf.org/doc/html/rfc6265#section-5.3 p6
+    host_only: bool
+    # https://datatracker.ietf.org/doc/html/rfc6265#section-5.3 p8
+    secure_only: bool
+    # https://datatracker.ietf.org/doc/html/rfc6265#section-5.3 p9
+    http_only: bool
+    # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite
+    same_site: Optional[str]
+
+    _natural_key_attrs = ["name", "value", "domain", "path"]
+
+    @classmethod
+    def format_reference_human_readable(cls, reference: Reference) -> str:
+        t = reference.tokenized
+
+        return f"{t.name} @ {t.domain.name}{t.path}"
+
+
+class RawCookie(OOI):
```

Comment:
Also I don't really understand the purpose of RawCookie. We already have the raw cookie in the HTTP header, what is the added value of duplicating that into a RawCookie instead of just parsing the Cookie from the HTTP header?

Reply:
It is mostly to decouple cookies from header so that users can add cookies from other sources, such as manually without having to add website -> resource -> header etc

Reply:
What is the use case for this? Because I can't really think of any.

Reply:
But that's a whole different concept than cookies that KAT observes on websites. I don't think it is a good idea to mix these two things and have such loggedin user cookies end up on the web interface and in reports etc. and users being able to use them to login to things they aren't supposed to be able to login to. It also don't think it is a good idea to duplicate all the cookie data just for this.

We should create a different request object for this and probably treat it like a secret. I also wonder how good this will work in practice given that sessions usually have a limited lifetime and might also be bound to a specific IP address.

Reply:
There's a functional difference between having a Cookie in a HTTP header, and a cookie from any other context.
An observed http header is just that, an http header. From there we extract a cookie in its raw form. A parsed cookie on the other hand has had operations done on it. The reason we do these operations is the same with how we parse Url's into complex objects. This logic should be part of a business rule, and not be part of the model because the model cannot hold turing complete code like Bits can.
Separating these two types also allows us to create an UnparseableCookie finding related to the rawcookie object if parsing fails. It also allows other sources of cookies to be added, (eg, user input, network traffic analysers, browser cookiejars).

Last, the concern about keeping live credentials in the graph makes sense, but is the only reasonable way to create session based logins from KAT. Doing this would require someone to configure the values somewhere as OOI's wich can then be mashed into session cookies by BIT logic (or form posts). These session details will always be present in KAT somewhere for this to function, and having them in the graph means we can invalidate them over time (as the cookie expires), and also stores them in a forensic way.

User Cookies as seen in the browser from some hapless user will not be ingested into Kat, unless there's a boefje specifically looking for them, in which case it would be logical to show them in the reports too. If any cleaning/anonymizing is to be done, it should be done in the boefje doing that gathering, or perhaps they should be stored as encrypted ooi's.

Reply:
A cookie is not a Httpheader. 
"set-cookie: $cookie" is a Httpheader.
"$cookie" is a cookie.
"Cookie: $cookie" is a Httpheader. 

Those are different things, and offer different use cases.

Cookies can be found in headers, and most likely are, but  might come from other places like the user. 
Cookies traveling towards the server have a `Cookie` header-name instead of a `set-cookie` header-name.

By not having the cookie parse logic in the http header parser we keep that parser cleaner, and have a better separation of concerns where The cookie parsing BIT only needs to parse rawcookies instead of all httpheaders. I think we should do the same for CSP headers, header comes in, we create an unparsed raw CSP object, and parse that into the CSP stanza's. This also makes sure that we have a logical place to store the parsing code outside of the Model, as the model wont be able to hold parsing code once the model becomes a declarative system.

Regarding response / requests:
Ideally we move to a response / request OOI model at some point, where  we also generate new possible requests from url's + cookies. However this would flood the graph with possibly short lived OOI's.
Since cookies travel both ways, with a request to the server, and with a response from the server, both would be reasonable references. However, since cookies are hostname + path bound, (which might include subdomains), I feel like a reference to the hostname is enough, including the direction would not make sense unless we also bind it to a request or response OOI and then use that direction to validate if the cookies would have survived this trip and be accepted by the browser based on hostname / path mentioned in the cookie itself, eg, find mismatches. 
I'd think this is for now beyond the scope of this cookie model. Whereas the proposed model would already give us lots of useful BIT logic to detect various (missing) flags, cookie-origins, suspicious third party vendors and the like.

Reply:
There is one issue remaining though.
Multiple cookies in a single set-cookie header. We are seeing conflicting results based on concatenation of the cookie string, or the combination of multiple similar named headers.

For Cookie headers: we see concatenation of cookies with a semicolon:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies

For http2 we see the same happening with the set-cookie values:
https://www.rfc-editor.org/rfc/rfc7540#section-8.1.2.5

For http header concatenation we see comma separation:
https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2

> Multiple message-header fields with the same field-name MAY be present in a message if and only if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)].

This seems to be reflected in other parsers:
https://github.com/nfriedly/set-cookie-parser/blob/master/test/set-cookie-parser.js

And also what python requests emits.

However the python simplecookie module does not understand the comma separation when the last argument of the cookie is just a keyword. Eg, Secure, instead of secure=True. (same is probably true for the Httponly keyword).

Separation on semicolon would make parsing hard to do with future keywords being added, as the only way to properly split multiple cookies is to recognize (or more specifically not recognize) the name=value and being a new cookie during parsing.

Reply:
`The section 4.1.1 of RFC 6265 defines the Secure attribute and specifies that it is a "cookie attribute" that can be present with no value or with the value true. If the Secure attribute is present without a value, it is treated as if the value is true.`

And

`A <cookie-value> can optionally be wrapped in double quotes and include any US-ASCII character excluding a control character, Whitespace, double quotes, comma, semicolon, and backslash.`

Cookie values 'should' also be encoded (base64)
Cookie Names are also not ascii or utf8 per-se. (but might be rendered as such when shown to a user)

Reply:
CookieName1=002230; path=/; Httponly; Secure, _CookieName2=lfsjndldjn; Path=/

Fails but

CookieName1=002230; path=/; Httponly; Secure; SameSite=Lax,  _CookieName2=lfsjndldjn; Path=/

Doesnt fail

Reply:
If we only include the response hostname in the rawcookie objects, we will be unable to report where a problematic cookie originated. For example if we have two paths, /foo and /bar, and /foo sends a good cookie with path=/, but /bar sends a cookie also with path=/ and but missing the secure flag and we have a bit that reports on that. Without the information from which url the cookie originated we can only report "on hostname.com there is a bad cookie somewhere" instead of "https://hostname.com/bar sends a bad cookie". I think the latter is a lot more useful to include in reports.